File: src/stylingdialog.cpp
#include "stylingdialog.h"
#include "desktopnotesapplet.h" // For applying style to applet

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFormLayout>
#include <QSlider>
#include <QPushButton>
#include <QSpinBox>
#include <QLabel>
#include <QColorDialog>
#include <QDialogButtonBox>
#include <QDebug>

StylingDialog::StylingDialog(DesktopNotesData::NoteStyle &currentStyle,
                               DesktopNotesApplet* targetApplet,
                               QWidget *parent)
    : QDialog(parent),
      m_currentStyleRef(currentStyle),
      m_originalStyle(currentStyle), // Make a copy for potential revert
      m_targetApplet(targetApplet)
{
    setWindowTitle(i18n("Note Styling"));
    setModal(true);

    setupUi();

    // Initialize controls with current style values
    m_transparencySlider->setValue(static_cast<int>(m_currentStyleRef.transparency * 100));
    m_transparencyValueLabel->setText(QString::number(m_currentStyleRef.transparency * 100) + "%");

    QColor initialColor(m_currentStyleRef.backgroundColor);
    m_colorPreviewLabel->setStyleSheet(QString("QLabel { background-color: %1; border: 1px solid black; }").arg(initialColor.name()));

    m_marginSlider->setValue(m_currentStyleRef.margin);
    m_marginSpinBox->setValue(m_currentStyleRef.margin);

    // Connections for real-time preview
    connect(m_transparencySlider, &QSlider::valueChanged, this, &StylingDialog::onTransparencyChanged);
    connect(m_marginSlider, &QSlider::valueChanged, this, &StylingDialog::onMarginChanged);
    connect(m_marginSpinBox, qOverload<int>(&QSpinBox::valueChanged), this, &StylingDialog::onMarginChanged); // For spinbox direct input
    connect(m_bgColorButton, &QPushButton::clicked, this, &StylingDialog::onChooseColorClicked);

    // Ok/Cancel buttons
    connect(m_okButton, &QPushButton::clicked, this, &StylingDialog::acceptDialog);
    connect(m_cancelButton, &QPushButton::clicked, this, &StylingDialog::rejectDialog);
}

StylingDialog::~StylingDialog()
{
    // Qt handles child widget deletion
}

void StylingDialog::setupUi()
{
    QVBoxLayout *mainLayout = new QVBoxLayout(this);
    QFormLayout *formLayout = new QFormLayout();

    // --- Transparency ---
    m_transparencyLabel = new QLabel(i18n("Transparency:"));
    m_transparencySlider = new QSlider(Qt::Horizontal);
    m_transparencySlider->setRange(0, 100); // 0% to 100%
    m_transparencyValueLabel = new QLabel(); // To show percentage
    QHBoxLayout *transparencyLayout = new QHBoxLayout();
    transparencyLayout->addWidget(m_transparencySlider);
    transparencyLayout->addWidget(m_transparencyValueLabel);
    formLayout->addRow(m_transparencyLabel, transparencyLayout);

    // --- Background Color ---
    m_bgColorLabel = new QLabel(i18n("Background Color:"));
    m_bgColorButton = new QPushButton(i18n("Choose Color..."));
    m_colorPreviewLabel = new QLabel();
    m_colorPreviewLabel->setMinimumSize(50, 20); // Ensure it's visible
    m_colorPreviewLabel->setAutoFillBackground(true);
    QHBoxLayout *bgColorLayout = new QHBoxLayout();
    bgColorLayout->addWidget(m_bgColorButton);
    bgColorLayout->addWidget(m_colorPreviewLabel);
    bgColorLayout->addStretch();
    formLayout->addRow(m_bgColorLabel, bgColorLayout);

    // --- Margin ---
    m_marginLabel = new QLabel(i18n("Margin (pixels):"));
    m_marginSlider = new QSlider(Qt::Horizontal);
    m_marginSlider->setRange(0, 50); // Example range for margin
    m_marginSpinBox = new QSpinBox();
    m_marginSpinBox->setRange(0, 50);
    QHBoxLayout *marginLayout = new QHBoxLayout();
    marginLayout->addWidget(m_marginSlider);
    marginLayout->addWidget(m_marginSpinBox);
    formLayout->addRow(m_marginLabel, marginLayout);

    // Sync slider and spinbox for margin
    connect(m_marginSlider, &QSlider::valueChanged, m_marginSpinBox, &QSpinBox::setValue);
    connect(m_marginSpinBox, qOverload<int>(&QSpinBox::valueChanged), m_marginSlider, &QSlider::setValue);


    mainLayout->addLayout(formLayout);

    // --- Ok/Cancel Buttons ---
    QDialogButtonBox *buttonBox = new QDialogButtonBox();
    m_okButton = buttonBox->addButton(QDialogButtonBox::Ok);
    m_cancelButton = buttonBox->addButton(QDialogButtonBox::Cancel);
    mainLayout->addWidget(buttonBox);

    setLayout(mainLayout);
}

void StylingDialog::onTransparencyChanged(int value)
{
    m_currentStyleRef.transparency = static_cast<double>(value) / 100.0;
    m_transparencyValueLabel->setText(QString::number(value) + "%");
    if (m_targetApplet) {
        applyStyleToApplet(m_currentStyleRef);
    }
}

void StylingDialog::onMarginChanged(int value)
{
    m_currentStyleRef.margin = value;
    // If slider and spinbox are connected, one will update the other.
    // Ensure the styleRef is updated from the primary signal source or both.
    // Here, both slider and spinbox valueChanged signals are connected to this slot.
    // To avoid double application if one updates the other which then calls this again,
    // check if the value actually changed.
    // However, Qt's signal/slot usually handles this well unless we create a loop.
    // For simplicity, assume it works.
    if (m_marginSlider->value() != value) m_marginSlider->setValue(value);
    if (m_marginSpinBox->value() != value) m_marginSpinBox->setValue(value);

    if (m_targetApplet) {
        applyStyleToApplet(m_currentStyleRef);
    }
}

void StylingDialog::onChooseColorClicked()
{
    QColor currentColor = QColor(m_currentStyleRef.backgroundColor);
    QColorDialog colorDialog(currentColor, this);
    // Enable alpha selection if we want to support RGBA hex codes in the future,
    // but spec says hex #RRGGBB. Transparency is separate.
    // colorDialog.setOption(QColorDialog::ShowAlphaChannel, true);

    if (colorDialog.exec() == QDialog::Accepted) {
        QColor newColor = colorDialog.selectedColor();
        if (newColor.isValid()) {
            m_currentStyleRef.backgroundColor = newColor.name(QColor::HexRgb); // #RRGGBB format
            m_colorPreviewLabel->setStyleSheet(QString("QLabel { background-color: %1; border: 1px solid black; }").arg(newColor.name()));
            if (m_targetApplet) {
                applyStyleToApplet(m_currentStyleRef);
            }
        }
    }
}

void StylingDialog::applyStyleToApplet(const DesktopNotesData::NoteStyle& styleToApply)
{
    if (m_targetApplet) {
        // The applet needs a public method to temporarily apply a style for preview
        // or directly manipulate its visual properties based on the style struct.
        // For now, let's assume DesktopNotesApplet has a method:
        // m_targetApplet->previewStyle(styleToApply);
        // Or, more directly, if m_currentStyleRef is the applet's actual style object:
        m_targetApplet->applyNoteStyle(); // This will use the (already updated) m_currentStyleRef
        m_targetApplet->update(); // Ensure repaint
    }
}

DesktopNotesData::NoteStyle StylingDialog::getAppliedStyle() const
{
    // This is called if OK is pressed. The m_currentStyleRef has been updated throughout.
    return m_currentStyleRef;
}

DesktopNotesData::NoteStyle StylingDialog::getOriginalStyle() const
{
    // This is used if Cancel is pressed.
    return m_originalStyle;
}


void StylingDialog::acceptDialog()
{
    // m_currentStyleRef has already been updated with all changes.
    // No further action needed on the style object itself.
    // The caller (DesktopNotesApplet) will take m_currentStyleRef and save it.
    QDialog::accept();
}

void StylingDialog::rejectDialog()
{
    // Revert the m_currentStyleRef back to m_originalStyle
    m_currentStyleRef = m_originalStyle;
    // And apply this reversion to the applet for visual feedback
    if (m_targetApplet) {
        applyStyleToApplet(m_currentStyleRef); // Apply the original style back
    }
    QDialog::reject();
}

// In DesktopNotesApplet.h, we would need to add:
// friend class StylingDialog; // If StylingDialog needs to access private members for preview
// or add a public method:
// void previewStyle(const DesktopNotesData::NoteStyle& style);

// In DesktopNotesApplet.cpp's handleStyling():
// StylingDialog dialog(m_note.style, this, this); // Pass 'this' as parent and targetApplet
// if (dialog.exec() == QDialog::Accepted) {
//     // m_note.style is already updated by reference.
//     applyNoteStyle(); // Ensure final application (might be redundant if preview did it well)
//     saveNoteData();
// } else {
//     // m_note.style was reverted by dialog's rejectDialog().
//     // applyNoteStyle(); // Ensure final reversion (might be redundant)
// }

// For applyNoteStyle in DesktopNotesApplet to work with the live changes,
// m_currentStyleRef in StylingDialog must indeed be a reference to m_note.style in the applet.
// Then, DesktopNotesApplet::applyNoteStyle() will always use the latest values from the dialog.
// DesktopNotesApplet::update() should also be called by applyNoteStyle().

File: src/styling_dialog.py
import sys
from PyQt6.QtWidgets import (
    QDialog, QVBoxLayout, QHBoxLayout, QLabel, QSlider, QPushButton,
    QColorDialog, QSpinBox, QDialogButtonBox, QApplication, QWidget
)
from PyQt6.QtGui import QColor, QPalette
from PyQt6.QtCore import Qt

# For testing, we need a way to simulate the main widget that this dialog controls
class MockNoteWidgetForStyling:
    def __init__(self, note_id, initial_style):
        self.note_id = note_id
        self.current_style = initial_style.copy() # Simulate current style of the widget
        self.widget_ref = QWidget() # A dummy QWidget to show color changes for testing dialog
        self.widget_ref.setAutoFillBackground(True)
        self.widget_ref.setMinimumSize(200,100)
        self.apply_style_to_mock(self.current_style)

    def update_style_preview(self, style_attribute, value):
        print(f"MockNoteWidget (ID {self.note_id}): Previewing style {style_attribute} = {value}")
        # In a real app, this would directly manipulate the NoteWidget's appearance
        temp_style = self.current_style.copy()
        temp_style[style_attribute] = value
        self.apply_style_to_mock(temp_style) # Apply to the internal mock for visual feedback

    def apply_style_to_mock(self, style_data):
        # This method is just for the dummy widget in the test.
        # Real widget would have its own styling update mechanism.
        bg_color = QColor(style_data.get("backgroundColor", "#FFFFFF"))
        transparency = style_data.get("transparency", 1.0)
        bg_color.setAlphaF(transparency)

        palette = self.widget_ref.palette()
        palette.setColor(QPalette.ColorRole.Window, bg_color)
        self.widget_ref.setPalette(palette)

        # Margin is harder to show on a simple QWidget without content.
        # We'd print it or have a nested widget.
        print(f"  Mock style applied: BG {bg_color.name()}, Alpha {transparency}, Margin {style_data.get('margin',0)}")


    def get_current_style_for_dialog(self) -> dict:
        # Returns the style in the format the dialog expects
        return self.current_style

    def save_style(self, style_data: dict):
        # This is where the real NoteWidget would persist the style (e.g., update DB)
        print(f"MockNoteWidget (ID {self.note_id}): 'Save' called with style: {style_data}")
        self.current_style = style_data.copy()
        self.apply_style_to_mock(self.current_style) # Ensure mock reflects saved state
        # In real app: update_note(self.note_id, {"style": self.current_style})
        #              self.note_qwidget_instance.refresh_display()

    def revert_style(self, style_data: dict):
        print(f"MockNoteWidget (ID {self.note_id}): 'Revert' called with style: {style_data}")
        self.current_style = style_data.copy()
        self.apply_style_to_mock(self.current_style)
        # In real app: update_note(self.note_id, {"style": self.current_style})
        #              self.note_qwidget_instance.refresh_display()


class StylingDialog(QDialog):
    def __init__(self, note_id: int, initial_style: dict, target_widget_interface, parent=None):
        super().__init__(parent)
        self.setWindowTitle(f"Style Note ID: {note_id}")
        self.note_id = note_id
        self.target_widget = target_widget_interface # Interface to update the actual NoteWidget

        self.initial_style = initial_style.copy() # Store style at dialog opening for Cancel
        self.current_preview_style = initial_style.copy() # For real-time updates

        self._init_ui()
        self._load_initial_style()

    def _init_ui(self):
        self.layout = QVBoxLayout(self)

        # Transparency
        self.layout.addWidget(QLabel("Transparency (0% = Opaque, 100% = Invisible):"))
        self.transparency_slider = QSlider(Qt.Orientation.Horizontal)
        self.transparency_slider.setRange(0, 100) # Representing 0.0 to 1.0 (inverted: 0 is opaque)
        self.transparency_slider.setToolTip("0% is fully opaque, 100% is fully transparent (invisible).")
        self.transparency_slider.valueChanged.connect(self._preview_transparency)
        self.layout.addWidget(self.transparency_slider)

        # Background Color
        bg_layout = QHBoxLayout()
        bg_layout.addWidget(QLabel("Background Color:"))
        self.bg_color_button = QPushButton("Choose Color")
        self.bg_color_button.clicked.connect(self._choose_bg_color)
        bg_layout.addWidget(self.bg_color_button)
        self.bg_color_preview = QLabel() # Shows the chosen color
        self.bg_color_preview.setFixedSize(50, 20)
        self.bg_color_preview.setAutoFillBackground(True)
        bg_layout.addWidget(self.bg_color_preview)
        self.layout.addLayout(bg_layout)

        # Margin
        margin_layout = QHBoxLayout()
        margin_layout.addWidget(QLabel("Margin (pixels):"))
        self.margin_spinbox = QSpinBox()
        self.margin_spinbox.setRange(0, 100) # Max margin 100px
        self.margin_spinbox.valueChanged.connect(self._preview_margin)
        # TODO: Add slider for margin as per spec? For now, spinbox is more precise.
        # self.margin_slider = QSlider(Qt.Orientation.Horizontal)
        # self.margin_slider.setRange(0,100)
        # self.margin_slider.valueChanged.connect(self.margin_spinbox.setValue)
        # self.margin_spinbox.valueChanged.connect(self.margin_slider.setValue)
        margin_layout.addWidget(self.margin_spinbox)
        self.layout.addLayout(margin_layout)

        # For testing the dialog standalone: show the mock widget
        if isinstance(self.target_widget, MockNoteWidgetForStyling):
            self.layout.addWidget(QLabel("Live Preview (Mock Widget):"))
            self.layout.addWidget(self.target_widget.widget_ref)


        # Ok and Cancel buttons
        self.button_box = QDialogButtonBox(QDialogButtonBox.StandardButton.Ok | QDialogButtonBox.StandardButton.Cancel)
        self.button_box.accepted.connect(self.accept)
        self.button_box.rejected.connect(self.reject)
        self.layout.addWidget(self.button_box)

        self.setLayout(self.layout)

    def _load_initial_style(self):
        # Transparency: Slider is 0 (opaque) to 100 (transparent). Style is 1.0 (opaque) to 0.0 (transparent)
        transparency_value = self.initial_style.get("transparency", 1.0)
        slider_value = int((1.0 - transparency_value) * 100)
        self.transparency_slider.setValue(slider_value)

        # Background Color
        bg_color_hex = self.initial_style.get("backgroundColor", "#FFFFFF")
        self._update_color_button_preview(QColor(bg_color_hex))

        # Margin
        margin_value = self.initial_style.get("margin", 5)
        self.margin_spinbox.setValue(margin_value)

    def _preview_transparency(self, slider_value: int):
        # Slider: 0 = opaque (1.0), 100 = transparent (0.0)
        transparency = 1.0 - (slider_value / 100.0)
        self.current_preview_style["transparency"] = transparency
        self.target_widget.update_style_preview("transparency", transparency)

    def _choose_bg_color(self):
        current_color = QColor(self.current_preview_style.get("backgroundColor", "#FFFFFF"))
        color = QColorDialog.getColor(current_color, self, "Select Background Color")
        if color.isValid():
            self.current_preview_style["backgroundColor"] = color.name() # #RRGGBB format
            self._update_color_button_preview(color)
            self.target_widget.update_style_preview("backgroundColor", color.name())

    def _update_color_button_preview(self, color: QColor):
        palette = self.bg_color_preview.palette()
        palette.setColor(QPalette.ColorRole.Window, color)
        self.bg_color_preview.setPalette(palette)

    def _preview_margin(self, value: int):
        self.current_preview_style["margin"] = value
        self.target_widget.update_style_preview("margin", value)

    def accept(self): # OK pressed
        # "Changes made ... are applied instantly ... Ok: Applies all current settings"
        # This implies the target_widget already reflects the previewed state.
        # So, we just need to ensure the database/final state is set to current_preview_style.
        self.target_widget.save_style(self.current_preview_style)
        super().accept()

    def reject(self): # Cancel pressed
        # Revert all styling changes made *during the current session*
        # back to what they were when the dialog was first opened.
        self.target_widget.revert_style(self.initial_style)
        super().reject()


if __name__ == '__main__':
    app = QApplication(sys.argv)

    # Initial style for the mock widget
    mock_initial_style = {
        "transparency": 0.8, # 80% opaque
        "backgroundColor": "#E0FFFF", # Light cyan
        "margin": 10
    }

    # Create the mock widget that the dialog will control
    mock_widget_controller = MockNoteWidgetForStyling(note_id=1, initial_style=mock_initial_style)

    dialog = StylingDialog(
        note_id=1,
        initial_style=mock_widget_controller.get_current_style_for_dialog(),
        target_widget_interface=mock_widget_controller
    )

    # Show the mock widget itself (not part of dialog, but dialog updates it)
    # In real app, this would be the actual NoteWidget on desktop
    # For test, let's put it in a simple window
    test_host_window = QWidget()
    test_layout = QVBoxLayout(test_host_window)
    test_layout.addWidget(QLabel("This is the 'actual' Note Widget (Mocked for Test):"))
    test_layout.addWidget(mock_widget_controller.widget_ref) # Add the QWidget part of the mock
    test_host_window.setWindowTitle("Mock Note Widget Host")
    test_host_window.show()

    if dialog.exec(): # Show the dialog modally
        print("Styling Dialog accepted (OK).")
        print("Final style on mock widget:", mock_widget_controller.current_style)
    else:
        print("Styling Dialog rejected (Cancel).")
        print("Style on mock widget after cancel:", mock_widget_controller.current_style)
        # Verify it reverted to initial_style
        assert mock_widget_controller.current_style == mock_initial_style
        print("Style successfully reverted to initial state.")

    sys.exit(app.exec())
